rust/sbp/src/sbp_string.rs

```diff
@@ -37,6 +38,94 @@ where
     }
 }
 
+impl<T: AsRef<[u8]>> SbpString<T, Unterminated> {
+    /// Checked unterminated SbpString builder, verify last byte is non null
+    pub fn unterminated(data: T) -> Result<Self, UnterminatedError> {
+        match data.as_ref().last() {
+            Some(l) if l != &0 => Ok(Self::new(data)),
+            _ => Err(UnterminatedError),
+        }
+    }
+}
+
+impl<T: AsRef<[u8]>> SbpString<T, NullTerminated> {
+    pub fn null_terminated(data: T) -> Result<Self, NullTerminatedError> {
+        match data.as_ref().last() {
+            Some(l) if l == &0 => Ok(Self::new(data)),
+            _ => Err(NullTerminatedError),
+        }
+    }
+}
```

Comment:
What do you think about having separate constructors for Vec<u8> and [u8; LEN] strings? In the case of the Vec I think it might be nice to have the constructor be infallible. It's a variable length string so we can just add a null byte. And for the fixed strings we could still return an error (in addition to now checking the length of the string (at compile time)). Or we could make everything infallible by padding short strings + setting the last byte to 0  

```rust
impl SbpString<Vec<u8>, NullTerminated> {
    pub fn null_terminated(data: impl Into<Vec<u8>>) -> Self {
        let mut data = data.into();
        if data.last().map_or(true, |n| n != &0) {
            data.push(0);
        }
        Self::new(data)
    }
}

impl<const LEN: usize> SbpString<[u8; LEN], NullTerminated> {
    // accepting an array instead of AsRef<[u8]> checks the length at compile time
    pub fn null_terminated(data: [u8; LEN]) -> Result<Self, NullTerminatedError> {
        match data.last() {
            Some(l) if l == &0 => Ok(Self::new(data)),
            _ => Err(NullTerminatedError),
        }
    }
}
// or if we want an infallible api
impl<const LEN: usize> SbpString<[u8; LEN], NullTerminated> {
    pub fn null_terminated(data: impl AsRef<[u8]>) -> Self {
        let mut s: [u8; LEN] = [0; LEN];
        s.copy_from_slice(data.as_ref());
        s[LEN - 1] = 0;
        Self::new(s)
    }
}
```

what do you think, too magic?

Reply:
hmmm aren't both of these the same?
i.e. `[u8; LEN]` is also just `impl Into<Vec<u8>>`

running
`SbpString::null_terminated([0x1, 0x1]);` would use both implementation (as before test case worked for into vec, but now with [u8;2] as well)